### PR TITLE
Xml documentation utilities to automatically remove unshown nodes

### PR DESCRIPTION
**Pull request by Sancerio**
Hi @Benglin  and @sharadkjaiswal , please review and merge.

The changes are:
- CreateInstallers.bat is updated to run this program before it moves localization resource assemblies

- New project is added under src/Tools/XmlDocumentationsUtility which mainly contains program.cs as it main program and a ZeroTouchModule class

- Test automation is created under this directory as well

- program.cs searches en-US directory recursively inside the bin folder

- only xml files inside en-US need to be modified since other language resources only follow the en-US resources

- program.cs basically extracts method, type, property,and field from the library's xml documentation files and pass them to the ZeroTouchModule that checks whether they are shown in the UI or not.

### CODE REVIEW

File: src/Tools/XmlDocumentationsUtility/Program.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+{
+    public class XmlDocumentationsUtility
+    {
+
+        static void Main(string[] args)
+        {
+            try
+            {
+                //remove and check for node documentation xml under Debug configuration
+                string xmlDir = Path.Combine(args[0], "bin", "AnyCPU", "Debug", "en-US");
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Program.cs]: We shouldn't hardcode the path as Debug or Release. arg[0] should point to Dynamo bin location.

File: src/Tools/XmlDocumentationsUtility/Program.cs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                }
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine("The process failed: {0}", e.ToString());
+            }
+            
+        }
+
+        private static void CheckAndDeleteXmlNodes(string xmlPath,ZeroTouchModule zeroTouchModule)
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Program.cs]: RemoveDocumentationForHiddenNodes()

File: src/Tools/XmlDocumentationsUtility/Program.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            xml.Load(xmlPath);        
+            XmlNodeList elemList = xml.GetElementsByTagName("member");
+
+            for (int i = 0; i < elemList.Count; i++)
+            {
+                string attrVal = elemList[i].Attributes["name"].Value;
+                string type = "";
+                bool hasToBeRemoved = false;
+                int index = 0;
+                switch (attrVal[0])
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Program.cs]: MemberData ParseMemberElement(XmlElement element)
struct MemberData
{
enum Type;
string TypeName;
string MemberName;
}

File: src/Tools/XmlDocumentationsUtility/XmlDocumentationsUtility.csproj [hunk 4]
```diff
@@ -0,0 +1,69 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/XmlDocumentationsUtility.csproj]: Use config/CS.csprops

File: src/Tools/XmlDocumentationsUtility/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("XmlDocumentationsUtility")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Microsoft")]
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Properties/AssemblyInfo.cs]: Why Microsoft??

File: src/Dynamo.All.2013.sln [hunk 6]
```diff
@@ -187,6 +187,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DynamoCoreWpf", "DynamoCore
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DynamoServicesTests", "..\test\Libraries\NodeServicesTest\DynamoServicesTests.csproj", "{3E4278F7-60EC-491C-8094-B525C36A51A7}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "XmlDocumentationsUtility", "Tools\XmlDocumentationsUtility\XmlDocumentationsUtility.csproj", "{4CAB9005-0001-4B03-BBA0-57070ED105F1}"
```
Comment (lukechurch) [inline src/Dynamo.All.2013.sln]: Changes to 2013.SLN need to be copied into the 2012.SLN as well

**Comment by lukechurch** [comment]
Needs tests :)

**Comment by Benglin** [comment]
Thanks for the work, @Sancerio. Unfortunately I can't quite look at this now due to other priorities, but I trust that you have already learnt from the best.

**Comment by Sancerio** [comment]
@Benglin it's ok. Thanks for your help
@sharadkjaiswal: Please review and merge
@lukechurch : PTAL

File: src/Tools/XmlDocumentationsUtility/Test/ZeroTouchModuleTests.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        [SetUp]
+        public void SetUp()
+        {
+            module = new ZeroTouchModule("ProtoGeometry.dll");
+        }
+
+        [Test]
+        public void TypeExists()
+        {
+            Assert.IsFalse(module.TypeExists("Analysis.IAnalysisData"));
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Test/ZeroTouchModuleTests.cs]: Use some specific class which is present in ProtoGeometry but that is not available in ZeroTouchModule, such as HostFactory or any other class with [SuppressImportInVM] attribute.

**Comment by sharadkjaiswal** [comment]
LGTM

File: src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using XmlDocumentationsUtility;
+
+namespace XmlDocumentationsUtility.Test
+{
+    [TestFixture]
+    class XmlDocUtilityTests
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs]: Can you also add one more test for RemoveDocumentationForHiddenNodes() method and ensure that after removing hidden nodes, the new xml has all the nodes which was required.

Comment (Sancerio) [inline src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs]: Sure. Thanks for the feedback

**Comment by Sancerio** [comment]
done. @sharadkjaiswal , please review and merge 
@lukechurch  PTAL

**Comment by Sancerio** [comment]
@sharadkjaiswal Here is the attempt to test the RemoveDocumentationForHiddenNodes. Please review.

File: src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+using NodeDocumentationUtility;
+using System.IO;
+using System.Xml;
+
+namespace XmlDocumentationsUtility.Test
+{
+    [TestFixture]
+    class XmlDocUtilityTests
+    {
+        private ZeroTouchModule module = null;
```
Comment (sharadkjaiswal) [inline src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs]: we don't need this. module can be declared locally in the scope of the test.

**Comment by sharadkjaiswal** [comment]
Can you please create a separate PR for resource access modifier. Keep this pull request specific to Documentation utility.

File: src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs (hunk 9 above)
Comment (Sancerio) [inline src/Tools/XmlDocumentationsUtility/Test/XmlDocUtilityTests.cs]: Ah, ok. please see the latest commit

**Comment by Sancerio** [comment]
Sorry for that, will seperate the things for now

**Comment by Sancerio** [comment]
@sharadkjaiswal done, many thanks for your help!

**Comment by sharadkjaiswal** [comment]
Thanks for you work @Sancerio LGTM!

**Pull request merged**